lib/Utils/Logger.cpp

```diff
@@ -51,74 +47,166 @@ std::string toString(const SeverityLevel &Level) {
   }
 }
 
-SeverityLevel toSeverityLevel(const std::string &S) {
-  SeverityLevel Type = llvm::StringSwitch<SeverityLevel>(S)
-#define SEVERITY_LEVEL(NAME, TYPE) .Case(NAME, SeverityLevel::TYPE)
-#include "phasar/Utils/SeverityLevel.def"
-                           .Default(SeverityLevel::INVALID);
-  return Type;
+void Logger::setLoggerFilterLevel(SeverityLevel Level) {
+  LogFilterLevel = Level;
 }
 
-std::ostream &operator<<(std::ostream &OS, const SeverityLevel &Level) {
-  return OS << toString(Level);
+SeverityLevel Logger::getLoggerFilterLevel() { return LogFilterLevel; }
+
+bool Logger::isLoggingEnabled() { return LoggingEnabled; }
+
+void Logger::initializeStdoutLogger(
+    const std::optional<SeverityLevel> &Level,
+    const std::optional<std::string> &Category) {
+  LoggingEnabled = true;
+  if (Category.has_value()) {
+    CategoriesToStreamVariant[Category.value()][Level] = StdStream::STDOUT;
+  } else {
+    LevelsToStreamVariant[Level] = StdStream::STDOUT;
+  }
 }
 
-void setLoggerFilterLevel(SeverityLevel Level) { LogFilterLevel = Level; }
+void Logger::initializeStderrLogger(
+    const std::optional<SeverityLevel> &Level,
+    const std::optional<std::string> &Category) {
+  LoggingEnabled = true;
+  if (Category.has_value()) {
+    CategoriesToStreamVariant[Category.value()][Level] = StdStream::STDERR;
+  } else {
+    LevelsToStreamVariant[Level] = StdStream::STDERR;
+  }
+}
 
-bool logFilter(const boost::log::attribute_value_set &Set) {
-#ifdef DYNAMIC_LOG
-  return Set["Severity"].extract<SeverityLevel>() >= LogFilterLevel;
-#else
-  return false;
-#endif
+[[nodiscard]] bool Logger::initializeFileLogger(
+    const llvm::StringRef &Filename, const std::optional<SeverityLevel> &Level,
+    const std::optional<std::string> &Category, bool Append) {
+  LoggingEnabled = true;
+  if (Category.has_value()) {
+    CategoriesToStreamVariant[Category.value()][Level] = Filename.str();
+  } else {
+    LevelsToStreamVariant[Level] = Filename.str();
+  }
+  if (LogfileStreams[Filename] == nullptr) {
+    std::error_code EC;
+    if (Append) {
+      LogfileStreams[Filename] = std::make_unique<llvm::raw_fd_ostream>(
+          Filename, EC,
+          llvm::sys::fs::OpenFlags::OF_Append |
+              llvm::sys::fs::OpenFlags::OF_ChildInherit);
+    } else {
+      LogfileStreams[Filename] = std::make_unique<llvm::raw_fd_ostream>(
+          Filename, EC, llvm::sys::fs::OpenFlags::OF_ChildInherit);
+    }
+    // Following
+    // https://stackoverflow.com/questions/41699343/how-do-i-test-that-an-stderror-code-is-not-an-error
+    if (EC) {
+      LogfileStreams[Filename] = nullptr;
+      llvm::errs() << "Failed to open logfile: " << Filename << '\n';
+      llvm::errs() << EC.message() << '\n';
+      return false;
+    }
```

Comment:
```suggestion

  std::error_code EC;
  auto [It, Inserted] = [&] {
    if (Append) {
      return LogfileStreams.try_emplace(
          Filename, Filename, EC,
          llvm::sys::fs::OpenFlags::OF_Append |
              llvm::sys::fs::OpenFlags::OF_ChildInherit);
    }

    return LogfileStreams.try_emplace(
        Filename, Filename, EC, llvm::sys::fs::OpenFlags::OF_ChildInherit);
  }();

  if (!Inserted) {
    return true;
  }

  if (EC) {
    LogfileStreams.erase(Filename);
    llvm::errs() << "Failed to open logfile: " << Filename << '\n';
    llvm::errs() << EC.message() << '\n';
    return false;
  }
```

To compensate the missing `unique_ptr` indirection from the comment in the header

Reply:
What happens if the stream cannot be opened? then no insertion takes place and true is returned ...?